### PR TITLE
Harden compact leaf-log handoff cleanup

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Refs #3050. Harden exhaustive compact leaf-page-log handoff cleanup so failed or interrupted runs cannot silently leave stale active writer state or report success after cleanup fails.

## Context

This is speculative downstream work on top of draft PR #3069 / issue #3049. Keep this PR draft until #3049 merges, then rebase onto `main` and rerun validation.

## Non-goals

- No metadata format or recovery format changes.
- No cached wrapper checkpoint/close/flush behavior changes.
- No benchmark or performance tuning changes.
- No AI reviews requested while this remains a speculative downstream draft.

## Design

- Adds typed `ErrCompactStorageLeafPageLogHandoffCleanup` / `CompactStorageLeafPageLogHandoffError` for cleanup/restore failures after CompactStorage installs its temporary compact writer.
- Makes compact handoff cleanup explicit and idempotent.
- Restores the exact previous leaf-page-log owner only after advancing its compact handoff sequence past compact-created leaf segments.
- If the previous owner cannot be advanced or segment scanning fails, clears the active leaf-page-log writer so the DB fails closed and callers must close/reopen before resuming writes.
- Joins cleanup errors into interrupted compact errors instead of swallowing them.
- Delays `FullyCompacted` / `PolicyFullyCompacted` / `ByteMinimized` success flags until after handoff cleanup succeeds.

## Scope

- Includes:
  - `TreeDB/db/compact_storage.go`
  - `TreeDB/db/compact_storage_test.go`
  - `TreeDB/db/compact_storage_owner_matrix_test.go`
  - `TreeDB/compact_storage.go`
- Excludes:
  - On-disk/recovery format changes
  - Cached wrapper maintenance implementation
  - CLI/reporting changes

## Correctness

- Tests run:
  - `GOWORK=off go test ./TreeDB/db -run 'TestCompactStorage(LeafPageLogHandoff|PreInstall|Interrupted|ExhaustiveRefusesExternal|RawBackendCommandWAL)'`
  - `GOWORK=off go test ./TreeDB/db -run 'CompactStorage|LeafPageLog|Reopen|Corrupt|Truncate|Recovery'`
  - `GOWORK=off go test ./TreeDB -run 'Reopen|Recovery|Compact'`
  - `git diff --check`
- Corruption/edge cases covered:
  - Pre-install RID scan failure leaves previous owner/data writable and readable.
  - Interrupted compact after install/during rewrite restores exact previous owner and survives close/reopen.
  - Interrupted compact after value-log rewrite checkpoint restores exact previous owner and survives close/reopen.
  - Interrupted compact after leaf-generation seal restores exact previous owner and survives close/reopen.
  - Previous-owner advance failure returns typed cleanup error and clears active writer fail-closed.
  - External unsupported leaf-page-log owner remains typed, non-mutating, and returns no success flags.
- Concurrency/race coverage:
  - No new concurrent state machine. Existing maintenance lock remains the mutation boundary.

## Performance

- Benchmarks run: none.
- Reason: cleanup/status hardening only; no hot-path or data-structure change intended.

## Rollout / Toggle Plan

- Default setting: always-on cleanup/status hardening.
- How to disable quickly: revert this PR.

## Follow-ups

- Keep draft until #3049 / PR #3069 merges.
- Rebase onto `main` after #3049 lands and rerun the required validation before marking ready.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved CompactStorage reliability when handing off and restoring cached leaf-page logs.
  * Added safer fail-closed handling for cleanup and recovery errors during compaction.
  * Preserved existing leaf-page-log ownership when compaction cannot complete.
  * Refined classification of cached versus external leaf-page-log setups, reducing incorrect refusals.
  * Expanded test coverage for interruption, recovery, and error reporting scenarios.

### CODE REVIEW

**Comment by snissn** [comment]
Coordinator update for head `2213e5e692f879685c80eb677f6c747a2d8240f5`:

This head preserves the existing PR branch history, merges latest `origin/main` `635496ece98a891a6417d5ec69d5a08ae461a16e`, and retargets the PR to `main` after #3049 merged.

Validation on the actual pushed merge-update branch `/tmp/gomap-3050-merge-update`:
- `git diff --check`
- `GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffRestoresPreviousAndAdvancesSeq|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStoragePreInstallFailureDoesNotMutateLeafPageLogOwner|TestCompactStorageInterruptedHandoffRestoresPreviousLeafPageLog|TestCompactStorageRawBackendCommandWALValueLogLeafPageLogCurrentSupported|TestCompactStorageActiveWriterLifecycleIsModeledStatus|TestCompactStorageDefaultProducedOwnerContractMatrix|TestCompactStorageLeafPageLogOwnerMatrix'`
- `GOWORK=off go test ./TreeDB ./TreeDB/db -count=1 -run 'TestCompactStoragePublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageActiveWriterLifecycleIsModeledStatus|TestCompactStorageDefaultProducedOwnerContractMatrix|TestCompactStorageLeafPageLogHandoffRestoresPreviousAndAdvancesSeq|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStoragePreInstallFailureDoesNotMutateLeafPageLogOwner|TestCompactStorageInterruptedHandoffRestoresPreviousLeafPageLog|TestCompactStorageRawBackendCommandWALValueLogLeafPageLogCurrentSupported|TestCompactStorage'`
- `GOWORK=off go test ./TreeDB/caching -count=1 -run 'LeafPageLog|Compact|Checkpoint|Close|Flush|ValueLog'`
- after the final clean merge to `635496ece`: `git diff --check`, the focused `./TreeDB/db` compact-storage slice, and the focused `./TreeDB/caching` slice were rerun and passed.

Merge gate remains blocked until latest-head CI is green and fresh review/thread disposition is clean.

**Comment by snissn** [comment]
@codex review
@coderabbitai review
@copilot review

**Comment by snissn** [comment]
Coordinator update for head `4f842969b`:

Tightened the #3050 production-closeout evidence after an independent internal review found residual coverage gaps:
- Replaced cached-owner shape inference with an explicit cached-wrapper owner marker implemented by the caching leaf-log types, so caller-owned logs with similar concurrency/segment-provider capabilities stay classified as external unsupported.
- Expanded the public cached exhaustive compact refusal test to assert the quiescence contract names background flush/apply workers, checkpoint/close drains, and cached backlog.
- Added deterministic fail-closed cleanup coverage for compact writer close failures and compact leaf-segment scan failures, alongside the existing restore/advance cleanup failure coverage.

Validation on `/tmp/gomap-3050-merge-update`:
- `git diff --check`
- `GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageActiveWriterLifecycleIsModeledStatus|TestCompactStorageDefaultProducedOwnerContractMatrix'`
- `GOWORK=off go test ./TreeDB -count=1 -run 'TestCompactStoragePublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper'`
- `GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStoragePublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageActiveWriterLifecycleIsModeledStatus|TestCompactStorageDefaultProducedOwnerContractMatrix|TestCompactStorageLeafPageLogHandoffRestoresPreviousAndAdvancesSeq|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed|TestCompactStoragePreInstallFailureDoesNotMutateLeafPageLogOwner|TestCompactStorageInterruptedHandoffRestoresPreviousLeafPageLog|TestCompactStorageRawBackendCommandWALValueLogLeafPageLogCurrentSupported|TestCompactStorage'`
- `GOWORK=off go test ./TreeDB ./TreeDB/db -count=1 -run 'TestCompactStoragePublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageActiveWriterLifecycleIsModeledStatus|TestCompactStorageDefaultProducedOwnerContractMatrix|TestCompactStorageLeafPageLogHandoffRestoresPreviousAndAdvancesSeq|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed|TestCompactStoragePreInstallFailureDoesNotMutateLeafPageLogOwner|TestCompactStorageInterruptedHandoffRestoresPreviousLeafPageLog|TestCompactStorageRawBackendCommandWALValueLogLeafPageLogCurrentSupported|TestCompactStorage'`
- `GOWORK=off go test ./TreeDB/db -count=1 -run 'CompactStorage|LeafPageLog|ValueLog|Reopen'`
- `GOWORK=off go test ./TreeDB/caching -count=1 -run 'LeafPageLog|Compact|Checkpoint|Close|Flush|ValueLog'`

Merge gate remains blocked until latest-head CI is green and fresh review/thread disposition is clean.

**Comment by snissn** [comment]
@codex review
@coderabbitai review
@copilot review

**Comment by snissn** [comment]
Coordinator update for head `cdd40e8969a2db61e8dc026e7e04a872d3220db1`:

- Fixed the independent closeout review finding: compact leaf-page-log handoff cleanup now fails closed immediately on compact-writer close errors, clears the active writer, and does not restore/advance a previous owner after a close-stage failure.
- Extended the close-failure regression test to install a replaceable previous owner and assert it is not restored or advanced after the injected close error.
- Local validation passed:
  - `git diff --check`
  - `GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce'`
  - `GOWORK=off go test ./TreeDB/db ./TreeDB -count=1 -run 'TestCompactStorage(LeafPageLog|PublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|RawBackendCommandWALValueLogLeafPageLogCurrentSupported|ActiveWriterLifecycleIsModeledStatus|DefaultProducedOwnerContractMatrix)'`
  - `GOWORK=off go test ./TreeDB/caching -count=1 -run 'LeafPageLog|Compact|Checkpoint|Close|Flush|ValueLog'`
  - `GOWORK=off go test ./TreeDB/db -count=1`

CI is now expected to rerun on this pushed head before merge consideration.

**Comment by snissn** [comment]
@codex review
@coderabbitai review
@copilot review

File: TreeDB/db/compact_storage.go [hunk 1]
```diff
@@ -1402,17 +1448,66 @@ func compactStorageValueLogFileID(name string) (uint32, bool) {
...
+	if err != nil {
+		h.db.setLeafPageLogRaw(nil)
+		return errors.Join(cleanupErr, compactStorageLeafPageLogHandoffError("scan compact leaf segments", err))
+	}
+	leafSeq := maxRewriteLaneSeq(segments, rewriteLeafLogLaneID)
+	if err := compactStorageAdvanceLeafPageLogSeqAtLeast(h.previousLeafPageLog, leafSeq); err != nil {
+		h.db.setLeafPageLogRaw(nil)
+		return errors.Join(cleanupErr, compactStorageLeafPageLogHandoffError("restore previous owner", err))
+	}
+	h.db.setLeafPageLogRaw(h.previousLeafPageLog)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/compact_storage.go:1492]: Fixed in cdd40e8969a2db61e8dc026e7e04a872d3220db1: cleanup now clears the active writer and returns immediately on compact-writer close failure, and the regression test covers the previous-owner branch.

**Comment by snissn** [comment]
Coordinator review-thread disposition for head `cdd40e8969a2db61e8dc026e7e04a872d3220db1`:

- The close-stage fail-closed finding is fixed in this head: cleanup clears the active leaf-page log and returns immediately after a compact-writer close error, so a previous owner cannot be restored after an uncertain close.
- `TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed` now installs a replaceable previous owner and asserts the previous owner is neither restored nor advanced on the injected close failure.

I am resolving the automated review thread for this now-fixed finding. New latest-head review findings, if any, should remain blockers until addressed.

**Comment by snissn** [comment]
Coordinator latest-base update for head `cf7c72b1ca34f3b9ae56f4c7e4646a5ae8a242df`:

- Merged latest `origin/main` `60cf7d22d5ea34b794d172f2f353e42572dbe9e9` into the PR branch without conflicts.
- Local validation passed after the latest-base merge:
  - `git diff --check`
  - `GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageDefaultProducedOwnerContractMatrix'`
  - `GOWORK=off go test ./TreeDB/db ./TreeDB -count=1 -run 'TestCompactStorage(LeafPageLog|PublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|RawBackendCommandWALValueLogLeafPageLogCurrentSupported|ActiveWriterLifecycleIsModeledStatus|DefaultProducedOwnerContractMatrix)'`
  - `GOWORK=off go test ./TreeDB/caching -count=1 -run 'LeafPageLog|Compact|Checkpoint|Close|Flush|ValueLog'`
  - `GOWORK=off go test ./TreeDB/db -count=1`

Merge gate remains blocked until latest-head CI is green and review/thread disposition is clean on this head.

**Comment by snissn** [comment]
@codex review
@coderabbitai review
@copilot review

**Comment by snissn** [comment]
Coordinator update for head `9cc67532137f9fbf5b3f43bfeb2fbe67d04d50271`:

- Fixed the failing Windows CI check `windows-core-2` from run https://github.com/snissn/gomap/actions/runs/28223276250/job/83609196194.
- Root cause was test-only: `TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed` appended a compact leaf segment and then removed `leaf_vlog` while the compact leaf writer still held `value-l255-000001.log` open. Linux permits that unlink pattern; Windows rejects it with `The process cannot access the file because it is being used by another process`.
- The test now explicitly closes and clears the compact leaf writer before replacing the leaf-log directory, preserving the intended scan-failure path while avoiding a leaked Windows file handle.
- Merged latest `origin/main` `f4dfaf2f016823c95f2a4656abd5fb31493bd976` into this PR branch after #3022 landed.
- Local validation passed:
  - `git diff --check`
  - `GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed|TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageDefaultProducedOwnerContractMatrix'`
  - `GOWORK=off go test ./TreeDB/db ./TreeDB -count=1 -run 'TestCompactStorage(LeafPageLog|PublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|RawBackendCommandWALValueLogLeafPageLogCurrentSupported|ActiveWriterLifecycleIsModeledStatus|DefaultProducedOwnerContractMatrix)'`
  - `GOWORK=off go test ./TreeDB/caching -count=1 -run 'LeafPageLog|Compact|Checkpoint|Close|Flush|ValueLog'`
  - `GOWORK=off go test ./TreeDB/db -count=1`

Merge gate remains blocked until latest-head CI is green and review/thread disposition is clean on this head.

**Comment by snissn** [comment]
@codex review
@coderabbitai review
@copilot review

**Comment by snissn** [comment]
Follow-up on the second `windows-core-2` failure after `9cc6753`:

- Root cause: `TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed` was still relying on platform filesystem behavior by replacing `leaf_vlog` with a file. That produces the intended scan error on Linux, but did not force the cleanup scanner error on Windows.
- Fix: added a narrow package-level hook for the compact leaf handoff segment scan, mirroring the existing close hook, and changed this test to inject the scan error directly. Production behavior still uses `listValueLogSegments`.

Validation on `513fb1bc0`:

```sh
git diff --check
GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed'\nGOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed|TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageDefaultProducedOwnerContractMatrix'\nGOWORK=off go test ./TreeDB/db ./TreeDB -count=1 -run 'TestCompactStorage(LeafPageLog|PublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|RawBackendCommandWALValueLogLeafPageLogCurrentSupported|ActiveWriterLifecycleIsModeledStatus|DefaultProducedOwnerContractMatrix)'\nGOWORK=off go test ./TreeDB/caching -count=1 -run 'LeafPageLog|Compact|Checkpoint|Close|Flush|ValueLog'\nGOWORK=off go test ./TreeDB/db -count=1\n```\n\nRequesting fresh review on the new head.\n\n@codex review\n@coderabbitai review\n@copilot review

**Comment by snissn** [comment]
CodeRabbit close-failure disposition for latest head 513fb1bc0095d306c48f5a17fa9a502e39cc4485: the current cleanup path already fails closed at the compact writer close boundary. In `TreeDB/db/compact_storage.go`, `compactStorageLeafPageLogHandoffCloseWriter(h.writer)` errors are wrapped, `h.db.setLeafPageLogRaw(nil)` is called, and cleanup returns immediately before `compactStorageLeafPageLogHandoffListSegments`, `compactStorageAdvanceLeafPageLogSeqAtLeast`, or restoring `previousLeafPageLog` can run. The focused close-failure and scan-failure tests cover this behavior on the current head.

@coderabbitai review

**Comment by snissn** [comment]
Latest-main refresh for head d07f350589569e2668debb3707efc6074cc893a4:

Merged current `origin/main` / `80fa5c37fca8f43bda4989df4af68d55fd3af52a` into the #3050 branch. The incoming main change is typed-column collection setup work and did not conflict with compact handoff files.

Validation on the refreshed head:

```sh
git diff --check
GOWORK=off go test ./TreeDB/db -count=1 -run 'TestCompactStorageLeafPageLogHandoffScanFailureFailsClosed|TestCompactStorageLeafPageLogHandoffCloseFailureFailsClosed|TestCompactStorageLeafPageLogHandoffAdvanceFailureFailsClosed|TestCompactStorageLeafPageLogCleanupErrorPropagatesOnce|TestCompactStorageLeafPageLogOwnerMatrix|TestCompactStorageDefaultProducedOwnerContractMatrix'
GOWORK=off go test ./TreeDB/db ./TreeDB -count=1 -run 'TestCompactStorage(LeafPageLog|PublicCommandWALRelaxedExhaustiveFailsClosedForCachedWrapper|RawBackendCommandWALValueLogLeafPageLogCurrentSupported|ActiveWriterLifecycleIsModeledStatus|DefaultProducedOwnerContractMatrix)'
GOWORK=off go test ./TreeDB/caching -count=1 -run 'LeafPageLog|Compact|Checkpoint|Close|Flush|ValueLog'
GOWORK=off go test ./TreeDB/db -count=1
```

All passed locally. Latest-head GitHub CI is now expected to rerun on d07f3505.

@codex review
@coderabbitai review
@copilot review
